src/Tools/Source/RunTests/TestHistoryManager.cs

```diff
@@ -0,0 +1,213 @@
+        if (!TryGetEnvironmentVariable("SYSTEM_ACCESSTOKEN", out var accessToken)
+            || !TryGetEnvironmentVariable("SYSTEM_PHASENAME", out var phaseName))
+        {
+            Console.WriteLine("Missing required environment variables - skipping test history lookup");
+            return ImmutableDictionary<string, TimeSpan>.Empty;
+        }
+
+        // Use the target branch (in the case of PRs) or source branch to find the last successful build.
+        var targetBranch = Environment.GetEnvironmentVariable("SYSTEM_PULLREQUEST_TARGETBRANCH") ?? Environment.GetEnvironmentVariable("BUILD_SOURCEBRANCHNAME");
+        if (string.IsNullOrEmpty(targetBranch))
+        {
+            Console.WriteLine("Missing both PR target branch and build source branch environment variables - skipping test history lookup");
+            return ImmutableDictionary<string, TimeSpan>.Empty;
+        }
+
+        var credentials = new Microsoft.VisualStudio.Services.Common.VssBasicCredential(string.Empty, accessToken);
+
+        var connection = new VssConnection(s_projectUri, credentials);
+
+        using var buildClient = connection.GetClient<BuildHttpClient>();
+
+        Console.WriteLine($"Getting last successful build for branch {targetBranch}");
+        var adoBranch = $"refs/heads/{targetBranch}";
+        var lastSuccessfulBuild = await GetLastSuccessfulBuildAsync(RoslynCiBuildDefinitionId, adoBranch, buildClient, cancellationToken);
+        if (lastSuccessfulBuild == null)
+        {
+            // If this is a new branch we may not have any historical data for it.
+            ConsoleUtil.WriteLine($"Unable to get the last successful build for definition {RoslynCiBuildDefinitionId} and branch {targetBranch}");
+            return ImmutableDictionary<string, TimeSpan>.Empty;
+        }
+
+        using var testClient = connection.GetClient<TestResultsHttpClient>();
+        var runForThisStage = await GetRunForStageAsync(lastSuccessfulBuild, phaseName, testClient, cancellationToken);
+        if (runForThisStage == null)
+        {
+            // If this is a new stage, historical runs will not have any data for it.
+            ConsoleUtil.WriteLine($"Unable to get a run with name {phaseName} from build {lastSuccessfulBuild.Url}.");
+            return ImmutableDictionary<string, TimeSpan>.Empty;
+        }
+
+        ConsoleUtil.WriteLine($"Looking up test execution data for build {lastSuccessfulBuild.Id} on branch {targetBranch} and stage {phaseName}");
+
+        var totalTests = runForThisStage.TotalTests;
+
+        Dictionary<string, TimeSpan> testInfos = new();
+        var duplicateCount = 0;
+
+        // Get runtimes for all tests.
+        var timer = new Stopwatch();
+        timer.Start();
+        for (var i = 0; i < totalTests; i += MaxTestsReturnedPerRequest)
+        {
+            var testResults = await GetTestResultsAsync(runForThisStage, i, MaxTestsReturnedPerRequest, testClient, cancellationToken);
```

Comment:
It looks like this phase takes maybe 30 seconds on desktop. Should we tinker with firing the all the tasks off immediately, pushing them into an array or something, then awaiting them one by one?

There's no need to spend time on it in this PR, but thought we should keep it in our back pocket.

Reply:
Yup - we can play around with it (will do that separately).  We have to be careful not to run too many though that we get rate limited.